### PR TITLE
added tests for calculate_fee(), calculate_waste() and effective_value()

### PR DESCRIPTION
**Pull request by chippsmith**
I added tests: test_calculate_fee(), test_effective_value_when_less_than_zero(), test_effective_value_when_greater_than_zero(), test_effective_value when_less_than_zero(), test_calculate_waste_to_drain(), and test_calculate_waste_to_miner() per issue number 24.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -578,4 +578,97 @@ mod test {
         let result = select_coin_lowestlarger(&inputs, options);
         assert!(matches!(result, Err(SelectionError::InsufficientFunds)));
     }
+
+    #[test]
+    fn test_calculate_fee(){
+        let fee = calculate_fee(60, 5.0);
+        assert_eq!(fee, 300)
+    }
```
Comment (rajarshimaitra) [inline src/lib.rs:586]: While writing these type of arithmatic tests, its better if you go with some form of test vectors. Each case will test on various scenarios and more scenarios you can add more robust is your tests. Even for inputs with wrong data and assert that the functions can give the correct error message.

For example:

```
input: 60, 5, Output: 300.
input 60, 5.0, Output: 300.
input 60 , -5, Output: Invalid Fee rate
```
Etc. 

Same goes for other tests. You can start with a population of given inputs and expected outputs. And then the test just loops though each of them and assert that inputs matches with expected outputs.

Comment (chippsmith) [reply src/lib.rs:586]: Thank you for your suggestions, I'm working on this now.  Ive got my tests working for inputs that are weight: u32 and rate: f32 which is what calculate_fee() takes as arguments.  In your examples it looks like I should test for cases where rate is given as a u32 and if so I'm unsure how to test that since the types don't match, the code wont compile.  Also In order to fix the calculate_fee() function so it returns an error when fee rate is negative I have to change the return type to a result and then call unwrap() everywhere in the code the function is called.  Am I heading in the right direction?

Comment (rajarshimaitra) [reply src/lib.rs]: Yes it's in the right direction. Don't take my example seriously, I didn't check the actual arg types while writing it. Its just for giving an idea.

Comment (rajarshimaitra) [reply src/lib.rs]: tests looks good. fmt is failing. Try to see if you can envision any more test edge cases. Or else this is good.

Comment (chippsmith) [reply src/lib.rs]: I will think about more test edge cases.  Fmt is format?  Where can I learn more about how this test works?  Is there a way to see where my code fails?

Comment (chippsmith) [reply src/lib.rs]: I added more edge cases and fixed the calculate_waste() and calculate_fee() functions to pass the new edge cases.  And I fixed the format.  Do you have any other suggestions?

File: src/lib.rs [hunk 2]
```diff
@@ -164,7 +171,7 @@ pub fn select_coin_lowestlarger(
         for (idx, input) in sorted_inputs.iter().skip(index) {
             accumulated_value += input.value;
             accumulated_weight += input.weight;
-            estimated_fees = calculate_fee(accumulated_weight, options.target_feerate);
+            estimated_fees = calculate_fee(accumulated_weight, options.target_feerate).unwrap();
```
Comment (rajarshimaitra) [inline src/lib.rs:174]: Ideally there should not be any unwraps in business code. All unwraps should be transformed into error propagation via `?`. I think calculate_fee can simple `?` as it returns the `FeeError`. Which should be included in the `SelectionError` enum.

File: src/lib.rs [hunk 3]
```diff
@@ -143,16 +150,16 @@ pub fn select_coin_lowestlarger(
     let target = options.target_value + options.min_drain_value;
 
     let mut sorted_inputs: Vec<_> = inputs.iter().enumerate().collect();
-    sorted_inputs.sort_by_key(|(_, input)| effective_value(input, options.target_feerate));
+    sorted_inputs.sort_by_key(|(_, input)| effective_value(input, options.target_feerate).unwrap());
```
Comment (rajarshimaitra) [inline src/lib.rs:153]: Same thing here. No unwraps. Just include the error variant in the SeelctionError.

File: src/lib.rs [hunk 4]
```diff
@@ -346,21 +355,29 @@ fn calculate_waste(
...
+    } else {
+        Ok((weight as f32 * rate).ceil() as u64)
+    }
 }
 
 /// Returns the effective value which is the actual value minus the estimated fee of the OutputGroup
 #[inline]
-fn effective_value(output: &OutputGroup, feerate: f32) -> u64 {
-    output
+fn effective_value(output: &OutputGroup, feerate: f32) -> Result<u64, FeeError> {
```
Comment (rajarshimaitra) [inline src/lib.rs:370]: Unrelated to this PR. But here it should be `long_term_fee_rate`? In that case, maybe makes sense to hardcode it at 10sats/vb like core? 

cc @delcin-raj

File: src/lib.rs [hunk 5]
```diff
@@ -346,21 +355,29 @@ fn calculate_waste(
...
+    Ok(output
         .value
-        .saturating_sub(calculate_fee(output.weight, feerate))
+        .saturating_sub(calculate_fee(output.weight, feerate)?))
```
Comment (rajarshimaitra) [inline src/lib.rs:373]: A better name for this could be `output.spending_weight`.  Weight can imply two things, the weight of the output, or the weight of the spending transaction's input. We need the second weight here.

**Review by rajarshimaitra — COMMENTED** [review]
Test vectors looks good.

Below few comments on unwraps, and few non-blocking comments, doesn't have to be addressed in the PR.

File: src/lib.rs (hunk 2 above)
Comment (chippsmith) [reply src/lib.rs:174]: Ive fixed this and will push changes once i fix the rest.

File: src/lib.rs (hunk 3 above)
Comment (rajarshimaitra) [reply src/lib.rs:149]: Ah alright. Ya, that can be done too. But not needed now. Maybe just put a comment here as a TODO so we don't forget?

Resolving this now.

File: src/lib.rs (hunk 5 above)
Comment (rajarshimaitra) [reply src/lib.rs:369]: Any input on this?

**Review by rajarshimaitra — APPROVED** [review]
utACK 4ebff2c6f09278a2af096296e53426731da50153

The test structure looks good to me. Although I haven't ran or checked the calculations.

One small nit remaining. 

ACK from me module @delcin-raj 's review.

File: src/lib.rs (hunk 5 above)
Comment (chippsmith) [reply src/lib.rs:369]: I agree that spending_weight is a better name than weight.  Perhaps input_weight or utxo_weight would be other options to consider.  Anything to differentiate the two would improve the readability of the code.

**Comment by chippsmith** [comment]
fmt failed because of an error in line 707.  It reads ```result: Ok(999_99_999_990)```, instead of ```result: Ok(99_999_999_990)```.  Im not sure if I should fix that or not since this is pending review from @delcin-raj.

**Comment by rajarshimaitra** [comment]
Should fix, unless the number itself is wrong..

File: src/lib.rs [hunk 6]
```diff
@@ -75,6 +76,8 @@ pub enum ExcessStrategy {
 pub enum SelectionError {
     InsufficientFunds,
     NoSolutionFound,
+    NegativeFeeRate,
```
Comment (delcin-raj) [inline src/lib.rs:79]: NonPositiveFeeRate will be accurate

File: src/lib.rs [hunk 7]
```diff
@@ -75,6 +76,8 @@ pub enum ExcessStrategy {
...
+    AbnormallyHighFee,
```
Comment (delcin-raj) [inline src/lib.rs:80]: AbnormallyHighFeeRate

**Review by delcin-raj — CHANGES_REQUESTED** [review]
I think these validations of low and high fee rate should be handled in `pub` functions.
Not in `calculate_fee`
